server/src/main/java/org/elasticsearch/cluster/coordination/JoinValidationService.java

```diff
@@ -229,27 +229,41 @@ private void processNextItem() {
         try {
             nextItem.run();
         } finally {
-            final var remaining = queueSize.decrementAndGet();
-            assert remaining >= 0;
-            if (remaining > 0) {
-                runProcessor();
+            var cleanupSuccess = false;
+            try {
+                final var remaining = queueSize.decrementAndGet();
+                assert remaining >= 0;
+                if (remaining > 0) {
+                    runProcessor();
+                }
+                cleanupSuccess = true;
+            } finally {
+                assert cleanupSuccess;
```

Comment:
The only way these blocks can exit erroneously is through an exception - so couldn't you just do

    } catch (Exception e) {
        assert false : e.getMessage()
        throw e;
    }

rather than defining a separate variable?

Reply:
That doesn't cover `Throwable` instances which do not inherit from `Exception`, and we prefer to avoid `catch (Throwable t)`.

Reply:
`assert` just results in an `AssertionError`, so by catching `Throwable` it's just turning one error into another. Catching `Exception` is turning an exception into an error, which is just as visible as `AssertionError`